### PR TITLE
Provide a way to exit handshaking early without an error.

### PR DESCRIPTION
**Pull request by markdroth**
This restores functionality that was accidentally removed in #8913.

### CODE REVIEW

File: src/core/ext/transport/chttp2/server/chttp2_server.c [hunk 1]
```diff
@@ -153,12 +153,21 @@ static void on_handshake_done(grpc_exec_ctx *exec_ctx, void *arg,
...
+      grpc_server_setup_transport(
+          exec_ctx, connection_state->server_state->server, transport,
+          connection_state->accepting_pollset, args->args);
+      grpc_chttp2_transport_start_reading(exec_ctx, transport,
+                                          args->read_buffer);
+    } else {
+      // If the handshaking succeeded but there is no endpoint, then the
+      // handshaker may have handed off the connection to some external
+      // code, so we can just clean up here without creating a transport.
+      grpc_slice_buffer_destroy(args->read_buffer);
```
Comment (yang-g) [inline src/core/ext/transport/chttp2/server/chttp2_server.c]: Shouldn't the handshaker setting early_exit handle these, as they may want to pass the buffer out together with the fd?

**Comment by yang-g** [comment]
Otherwise LGTM

File: src/core/ext/transport/chttp2/server/chttp2_server.c (hunk 1 above)
Comment (markdroth) [inline src/core/ext/transport/chttp2/server/chttp2_server.c]: Good idea.  In fact, for consistency, the handshaker should be responsible for cleaning up all of the args fields (endpoint, read buffer, and channel args).

Done.

**Comment by markdroth** [comment]
The interop test failure is a case of #8989.

No idea what's going on with the performance tests...

**Comment by markdroth** [comment]
The other remaining test failure here looks like #8389.

**Comment by markdroth** [comment]
Note: This is part of fixing #8744.

**Pull request merged**